hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/processor/StatefulFlatMapP.java

```diff
@@ -0,0 +1,157 @@
+    private final BiFunction<K, TimestampedItem<S>, TimestampedItem<S>> updateMapFn;
+    private final FlatMapper<T, OUT> flatMapper;
+
+    private final Map<K, TimestampedItem<S>> keyToState = new LruHashMap();
+    private long currentWm = Long.MIN_VALUE;
+    private Traverser<Entry<K, TimestampedItem<S>>> snapshotTraverser;
+
+    public StatefulFlatMapP(
+            long ttl,
+            @Nonnull Function<? super T, ? extends K> keyFn,
+            @Nonnull Supplier<? extends S> createFn,
+            @Nonnull BiFunction<? super S, ? super T, ? extends Traverser<R>> statefulFlatMapFn,
+            @Nonnull BiFunction<? super K, ? super R, ? extends OUT> mapToOutputFn
+    ) {
+        checkPositive(ttl, "TTL");
+        this.ttl = ttl;
+        this.keyFn = keyFn;
+        this.createFn = createFn;
+        this.statefulFlatMapFn = statefulFlatMapFn;
+        this.mapToOutputFn = mapToOutputFn;
+        this.updateMapFn = this::updateMap;
+        this.flatMapper = flatMapper(this::flatMapEvent);
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    protected boolean tryProcess(int ordinal, @Nonnull Object item) {
+        if (item instanceof JetEvent) {
+            JetEvent<T> jetEvent = (JetEvent<T>) item;
+            if (jetEvent.timestamp() < currentWm) {
+                logLateEvent(getLogger(), currentWm, item);
+                lazyIncrement(lateEventsDropped);
+                return true;
+            }
+        }
+        return flatMapper.tryProcess((T) item);
+    }
+
+    @Nonnull
+    private Traverser<OUT> flatMapEvent(T event) {
+        K key = keyFn.apply(event);
+        S state = keyToState.compute(key, updateMapFn).item();
+        return statefulFlatMapFn.apply(state, event)
+                                .map(r -> mapToOutputFn.apply(key, r));
+    }
+
+    @Nonnull
+    private TimestampedItem<S> updateMap(K ignored, TimestampedItem<S> tsAndState) {
+        if (tsAndState != null && currentWm != Long.MIN_VALUE) {
+            long lastTouched = tsAndState.timestamp();
+            if (lastTouched != Long.MIN_VALUE && lastTouched < currentWm - ttl) {
+                tsAndState = null;
+            }
+        }
+        return tsAndState != null
+                ? tsAndState.setTimestamp(currentWm)
+                : new TimestampedItem<S>(currentWm, createFn.get());
+    }
+
+    @Override
+    public boolean tryProcessWatermark(@Nonnull Watermark watermark) {
+        currentWm = watermark.timestamp();
+        assert currentWm != Long.MIN_VALUE : "Received WM == Long.MIN_VALUE";
+        long countEvicted = 0;
+        for (Iterator<Entry<K, TimestampedItem<S>>> iter = keyToState.entrySet().iterator(); iter.hasNext();) {
+            long timestamp = iter.next().getValue().timestamp();
+            if (timestamp == Long.MIN_VALUE || timestamp >= currentWm - ttl) {
+                break;
+            }
+            iter.remove();
+            if (++countEvicted == MAX_ITEMS_TO_EVICT) {
```

Comment:
What if you have a high throughput (of unique keys), but low number of watermarks? 
Could it be you have millions of items / s, but just 1 watermark per second?

Reply:
`LruHashMap` evicts the eldest item each time you put in a new one, and this mechanism acts on top of that to reduce the map's size when transitioning from a period of high activity to low activity.